src/image/image_type.hpp

```diff
@@ -53,6 +53,10 @@ class image_type {
   /// Construct from existing data
   image_type(const char* image_data, size_t height, size_t width,
              size_t channels, size_t image_data_size, int version, int format);
+  /// Construct from a Boost GIL Image
+  /// Note: image_type can be any of the Boost GIL image types.
+  template<typename image_type>
+  explicit image_type(const image_type &gil_image);
```

Comment:
Use a different name for the template typename parameter from the class name, E.g. BoostImage.  That's the reason for your compile failure. 

However, the problem is that this will catch anything, including non-boost types, and then fail later.  The syntax for restricting this is a little convoluted, but here it would be: 

```
template<typename T>
   explicit image_type(const T &gil_image, 
typename std::enable_if<
std::is_same<T, boost::gil::image_type_1>::value || std::is_same<T, boost::gil::image_type_2>::value || (more in here)>::type* = 0) { 
```

Reply:
That would require forward declaring the boost::gil image types, but turns out that was more complicated than we thought because as @znation pointed out, the boost includes don't seem resilient to having their own types forward declared.

Reply:
What's the error when that happens?  That's odd and should not be the case.  Something else is going on -- likely that the header isn't included where it's needed.

Reply:
Sent the error to you on a private thread. What if I gracefully handle passing in a non Boost GIL image type in the CPP?

Reply:
@ylow and I just discussed this and he recommended I include the boost/gil/typedefs.hpp header which is basically like a forward declaration. Also got rid of the template since @ylow said it would be cleaner to just have overloaded constructors. Let me know if this looks good!